File: backend/app/routers/graph.py
"""图谱查询路由（前端 P3）：实体列表、邻域、搜索。

返回前端 GraphData 结构：nodes [{id,name,type,documentId}] + edges [{source,target,type,confidence}]。
直接查 Neo4j Entity/RELATES，供图谱可视化视图消费。
"""

from fastapi import APIRouter, HTTPException, Query, Request

router = APIRouter(prefix="/api/graph", tags=["graph"])

_LIST_ENTITIES = """
MATCH (e:Entity)
RETURN e.entity_id AS entity_id, e.name AS name, e.entity_type AS entity_type,
       e.document_id AS document_id
ORDER BY e.name
LIMIT $limit
"""

_LIST_EDGES = """
MATCH (s:Entity)-[r:RELATES]->(t:Entity)
RETURN s.entity_id AS source, t.entity_id AS target, r.type AS type,
       r.confidence AS confidence
LIMIT $limit
"""

_NEIGHBORS = """
MATCH (center:Entity {entity_id: $entity_id})
OPTIONAL MATCH (center)-[r1]-(nbr:Entity)
WITH center, collect(DISTINCT nbr) AS neighbors
UNWIND CASE WHEN size(neighbors)=0 THEN [center] ELSE neighbors + [center] END AS n
WITH collect(DISTINCT n) AS all_nodes
UNWIND all_nodes AS node
OPTIONAL MATCH (a)-[r:RELATES]->(b)
  WHERE a IN all_nodes AND b IN all_nodes
RETURN collect(DISTINCT {
  entity_id: node.entity_id, name: node.name,
  entity_type: node.entity_type, document_id: node.document_id
}) AS nodes,
       collect(DISTINCT {
  source: startNode(r).entity_id, target: endNode(r).entity_id,
  type: r.type, confidence: r.confidence
}) AS edges
"""

_SEARCH = """
MATCH (e:Entity)
WHERE toLower(e.name) CONTAINS toLower($q) OR toLower(e.normalized_name) CONTAINS toLower($q)
RETURN e.entity_id AS entity_id, e.name AS name, e.entity_type AS entity_type,
       e.document_id AS document_id
ORDER BY e.name
LIMIT $limit
"""


def _node(row: dict) -> dict:
    return {
        "id": row["entity_id"],
        "name": row["name"] or "",
        "type": row["entity_type"] or "",
        "documentId": row["document_id"] or "",
    }


def _edge(row: dict) -> dict:
    return {
        "source": row["source"],
        "target": row["target"],
        "type": row["type"] or "",
        "confidence": row.get("confidence"),
    }


@router.get("/entities")
async def list_entities(
    request: Request, limit: int = Query(100, ge=1, le=1000)
) -> dict:
    """返回实体列表与它们之间的 RELATES 边（前端 GraphData）。"""
    driver = request.app.state.neo4j
    ent_records, _, _ = driver.execute_query(
        _LIST_ENTITIES, limit=limit, database_="neo4j"
    )
    edge_records, _, _ = driver.execute_query(
        _LIST_EDGES, limit=limit * 2, database_="neo4j"
    )
    entity_ids = {r["entity_id"] for r in ent_records}
    return {
        "nodes": [_node(r.data()) for r in ent_records],
        # 只保留两端都在当前实体集内的边，避免 limit 截断后出现悬空边
        "edges": [
            _edge(r.data())
            for r in edge_records
            if r["source"] in entity_ids and r["target"] in entity_ids
        ],
    }


@router.get("/entities/{entity_id}/neighbors")
async def get_neighbors(request: Request, entity_id: str) -> dict:
    """返回单个实体的 1 跳邻域（含中心节点）。"""
    driver = request.app.state.neo4j
    records, _, _ = driver.execute_query(
        _NEIGHBORS, entity_id=entity_id, database_="neo4j"
    )
    if not records or not records[0]["nodes"]:
        raise HTTPException(status_code=404, detail=f"实体不存在: {entity_id}")
    row = records[0]
    # 过滤 None（OPTIONAL MATCH 在无边时产生 null 边）
    edges = [e for e in row["edges"] if e and e.get("source") and e.get("target")]
    return {
        "nodes": [
            {
                "id": n["entity_id"],
                "name": n["name"] or "",
                "type": n["entity_type"] or "",
                "documentId": n["document_id"] or "",
            }
            for n in row["nodes"]
            if n and n.get("entity_id")
        ],
        "edges": [
            {
                "source": e["source"],
                "target": e["target"],
                "type": e["type"] or "",
                "confidence": e.get("confidence"),
            }
            for e in edges
        ],
    }


@router.get("/search")
async def search_entities(
    request: Request, q: str = Query(..., min_length=1), limit: int = Query(20, ge=1, le=100)
) -> list[dict]:
    """实体名称模糊搜索（name 或 normalized_name CONTAINS q，大小写不敏感）。"""
    driver = request.app.state.neo4j
    records, _, _ = driver.execute_query(
        _SEARCH, q=q, limit=limit, database_="neo4j"
    )
    return [_node(r.data()) for r in records]

File: frontend/src/api/graph.ts
import { apiFetch } from './client'
import type { GraphData, GraphEdge, GraphNode } from '../types'

/**
 * 图谱 API 领域层：后端调用 + 字段映射收口。
 *
 * 后端（routers/graph.py）返回的字段名与前端 types/graph.ts 不一致，
 * 全部映射收口在这里，View 层只认 GraphData / GraphNode / GraphEdge。
 *
 * 字段映射：
 * - node: 后端 {id, name, type, documentId} → 前端 {id, label:name, entityType:type}
 * - edge: 后端 {source, target, type, confidence} → 前端 {id:生成, source, target, relationType:type}
 *   （后端 edge 无 id，前端需 id 给 React key / Cytoscape element，用 source-target-type 生成）
 */

// 后端原始响应结构（仅本文件内部用，不导出）
interface RawNode {
  id: string
  name: string
  type: string
  documentId: string
}
interface RawEdge {
  source: string
  target: string
  type: string
  confidence?: number | null
}
interface RawGraph {
  nodes: RawNode[]
  edges: RawEdge[]
}

function mapNode(n: RawNode): GraphNode {
  return { id: n.id, label: n.name, entityType: n.type }
}

function mapEdge(e: RawEdge): GraphEdge {
  return {
    id: `${e.source}-${e.target}-${e.type}`,
    source: e.source,
    target: e.target,
    relationType: e.type,
  }
}

function mapGraph(raw: RawGraph): GraphData {
  return { nodes: raw.nodes.map(mapNode), edges: raw.edges.map(mapEdge) }
}

/** 加载全图：GET /api/graph/entities?limit=100 → {nodes,edges} */
export async function fetchGraph(limit = 100): Promise<GraphData> {
  const raw = await apiFetch<RawGraph>(`/api/graph/entities?limit=${limit}`)
  return mapGraph(raw)
}

/** 实体 1 跳邻域：GET /api/graph/entities/{id}/neighbors → {nodes,edges}（含中心） */
export async function fetchNeighbors(entityId: string): Promise<GraphData> {
  const raw = await apiFetch<RawGraph>(
    `/api/graph/entities/${encodeURIComponent(entityId)}/neighbors`,
  )
  return mapGraph(raw)
}

/** 实体名称模糊搜索：GET /api/graph/search?q=&limit=20 → [node]（只返点不返边） */
export async function searchEntities(q: string, limit = 20): Promise<GraphNode[]> {
  const raw = await apiFetch<RawNode[]>(
    `/api/graph/search?q=${encodeURIComponent(q)}&limit=${limit}`,
  )
  return raw.map(mapNode)
}
